implementations/rust/ockam/ockam_api/src/kafka/secure_channel_map.rs

```diff
@@ -0,0 +1,387 @@
+use ockam_identity::api::{
+    DecryptionRequest, DecryptionResponse, EncryptionRequest, EncryptionResponse,
+};
+use ockam_identity::authenticated_storage::AuthenticatedStorage;
+use ockam_identity::{Identity, IdentityVault, SecureChannelRegistryEntry, TrustEveryonePolicy};
+use ockam_node::compat::tokio::sync::Mutex;
+use ockam_node::Context;
+use serde::{Deserialize, Serialize};
+
+///Offer simple APIs to encrypt and decrypt kafka messages.
+/// Underneath it create secure channels for each topic/partition
+/// and uses them to encrypt the content.
+/// Multiple secure channels may be created for the same topic/partition
+/// but each will be explicitly label.
+/// It's the same for both producer and consumer although it could be split
+/// into two distinct implementation.
+///This is proxy trait to avoid propagating the vault implementation.
+#[async_trait]
+pub(crate) trait KafkaSecureChannelController: Send + Sync {
+    ///encrypt the content specifically for the consumer waiting for that topic name and partition
+    /// the the secure channel is missing it'll be created, this may take up to few seconds
+    /// the first time
+    async fn encrypt_content_for(
+        &self,
+        context: &mut Context,
+        topic_name: &str,
+        partition_id: i32,
+        content: Vec<u8>,
+    ) -> Result<(UniqueSecureChannelId, Vec<u8>), Error>;
+
+    ///decrypt the content based on the unique secure channel identifier
+    /// the secure channel is expected to be already initialized
+    async fn decrypt_content_for(
+        &self,
+        context: &mut Context,
+        secure_channel_id: UniqueSecureChannelId,
+        encrypted_content: Vec<u8>,
+    ) -> Result<Vec<u8>, Error>;
+
+    ///start forwarders in the orchestrator for each {partition}_{topic_name} combination
+    /// should be used only by the consumer.
+    /// does nothing if they were already created, but fails it they already exist.
+    async fn start_forwarders_for(
+        &self,
+        context: &mut Context,
+        topic_id: &str,
+        partitions: Vec<i32>,
+    ) -> Result<(), Error>;
+}
+
+pub(crate) struct KafkaSecureChannelControllerImpl<V: IdentityVault, S: AuthenticatedStorage> {
+    inner: Arc<Mutex<InnerSecureChannelControllerImpl<V, S>>>,
+}
+
+//had to manually implement since #[derive(Clone)] doesn't work well in this situation
+impl<V: IdentityVault, S: AuthenticatedStorage> Clone for KafkaSecureChannelControllerImpl<V, S> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+///Unique identifier for a specific secure_channel.
+/// Used in order to distinguish between secure channels created between
+/// the same identities.
+#[derive(Debug, Clone, Serialize, Deserialize, Message)]
+struct SecureChannelIdentifierMessage {
+    secure_channel_identifier: UniqueSecureChannelId,
+}
+
+pub(crate) type UniqueSecureChannelId = u64;
+type TopicPartition = (String, i32);
```

Comment:
I actually like the simplicity of the tuple here, and being different types there is no risk of getting them wrong.
Since the type is internal I don't think is worth making it a struct.